configcat/.net-sdk
# Config v6

### Describe the purpose of your pull request

Implements config v6 support.

Features/improvements:
* [x] Fetch new config json version (`config_v6.json`) and update config model
* [x] SDK key format validation (when client is not set up to use local-only flag overrides)
* [x] AND conditions
* [x] New condition types
   * Prerequisite flag condition
   * Segment condition
* [x] Changed comparison operators
   * (NOT) CONTAINS ANY OF
* [x] New comparison operators
   * BEFORE/AFTER
   * TEXT (NOT) EQUALS (both hashed and cleartext)
   * TEXT (NOT) STARTS/ENDS WITH ANY OF (both hashed and cleartext)
   * ARRAY CONTAINS ANY OF (both hashed and cleartext)
* [x] Percentage options within targeting rules
* [x] Percentage options based on custom attributes
* [x] Custom User Object attribute handling (don't force string values + autoconversions under the hood)
* [x] User Object attribute/comparison value trimming
* [x] Comparison value hashing changes (SHA256, salting)
* [x] Flag overrides vs. config json salt and segments
* [x] `EvaluationDetails.MatchedTargetingRule`/`MatchedPercentageOption` properties (rename + set combinations correctly)

Tests: 
* [x] Evaluation matrix tests
* [x] Evaluation log tests
* [x] SDK key format validation
* [x] Custom User Object attribute handling (don't force string values + autoconversions under the hood)
* [x] Flag overrides vs. prerequisite flags (circular dependency, comparison value type mismatch, dependent/prereq flag overriding)
* [x] Flag overrides vs. config json salt and segments
* [x] `EvaluationDetails.MatchedTargetingRule`/`MatchedPercentageOption` properties (rename + set combinations correctly)

Related improvements/fixes:
* [ ] ~We should go to the cache in all polling modes instead of using the in memory variable (see https://trello.com/c/rreKm64A)~ (done earlier)

Breaking changes:
- Rename the `MatchedEvaluationRule` property to `MatchedTargetingRule` and the `MatchedEvaluationPercentageRule` property to `MatchedPercentageOption` in `EvaluationDetails`.
- Change config model (`IConfig` and related interfaces/enums).
- Change nullability of `User.Custom` to disallow null values.
- Change the `User.AllAttributes` property into a method (`GetAllAttributes`) as it allocates under the hood.

### Requirement checklist (only if applicable)
